library/core/src/cell/lazy.rs

```diff
@@ -77,10 +83,47 @@ impl<T, F: FnOnce() -> T> LazyCell<T, F> {
     #[inline]
     #[unstable(feature = "once_cell", issue = "74465")]
     pub fn force(this: &LazyCell<T, F>) -> &T {
-        this.cell.get_or_init(|| match this.init.take() {
-            Some(f) => f(),
-            None => panic!("`Lazy` instance has previously been poisoned"),
-        })
+        let state = unsafe { &*this.state.get() };
+        match state {
+            State::Init(data) => data,
+            State::Uninit(_) => unsafe { LazyCell::really_init(this) },
+            State::Poisoned => panic!("LazyCell has previously been poisoned"),
+        }
+    }
+
+    /// # Safety
+    /// May only be called when the state is `Uninit`.
+    #[cold]
+    unsafe fn really_init(this: &LazyCell<T, F>) -> &T {
+        let state = unsafe { &mut *this.state.get() };
+        // Temporarily mark the state as poisoned. This prevents reentrant
+        // accesses and correctly poisons the cell if the closure panicked.
+        let State::Uninit(f) = mem::replace(state, State::Poisoned) else { unreachable!() };
+
+        let data = f();
+
+        // If the closure accessed the cell, the mutable borrow will be
+        // invalidated, so create a new one here.
+        let state = unsafe { &mut *this.state.get() };
+        *state = State::Init(data);
+
+        // A reference obtained by downcasting from the mutable borrow
+        // would become stale if other references are created in `force`.
+        // Borrow the state directly instead.
```

Comment:
```suggestion
        // A reference obtained by downcasting from the mutable borrow
        // would become stale the next time `force` is called (since there is a conflict
        // between the mutable reference here and the shared reference there).
        // Do a new shared borrow of the state instead.
```